### PR TITLE
Tenant Issuer Config Related Fixes

### LINKED ISSUE #607: Configuration to create Tenants with default permission values
As a Traction operator, I want a flag to control the issuer approval requirements, so that I can run an instance where Tenants can make themselves issuers without Innkeeper intervention
--

AC:
- There are flags for the Traction instance appropriate plugin that can be set in Helm and Docker
- These flags contain the "default permission values" for the endorser and public did permission settings
- When Tenant creation occurs, the Tenant **Config** sets the configured default for the connect to endorser and create public DID config values. 
IE these ones:
```
  "connect_to_endorsers": {
    "endorser_alias": " ... ",
    "ledger_id": " ... "
  },
  "create_public_did": [
    "string"
  ],
  ```
  can have default values in the plugin configuration.
- If the default values are not present or blank arrays, the tenant does not get those permissions set (and the Innkeeper must manually approve the permissions.
- The Tenant UI still behaves accordingly in the Issuance section

Relates to work previously done in
- https://github.com/bcgov/traction/issues/393
- https://github.com/bcgov/traction/issues/394
- https://github.com/bcgov/traction/issues/570

### LINKED ISSUE #608: Block endorser connection and public DID tenant endpoints if Innkeeper approval is not set
As a Traction operator I want the Innkeeper approval for issuance-setting to be enforced on the plugin endpoints, so that a Tenant can't call them directly to become an issuer.
--

This is handled in the Tenant UI now but needs to be blocked in the API endpoints as well

AC:
- These 2(?) endpoints check whether the Tenant calling them is allowed to call them based on the Innkeeper allowance config.
- https://traction-tenant-proxy-dev.apps.silver.devops.gov.bc.ca/api/doc#/traction-tenant/post_tenant_endorser_connection
- https://traction-tenant-proxy-dev.apps.silver.devops.gov.bc.ca/api/doc#/wallet/post_wallet_did_create
(I think, check these are the endpoints we want to affect when doing this ticket)
- The Tenant UI still works as before (likely no changes needed for the UI in this ticket)
- The API returns an appropriate response code and message telling the caller they can't do the operation

Related to previous work done in 
- https://github.com/bcgov/traction/issues/393
- https://github.com/bcgov/traction/issues/394

### LINKED ISSUE #667: Can't clear the connect_to_endorser and create_public_did fields in config
In the 
PUT /innkeeper/tenants/{tenant-id}/config
https://traction-tenant-proxy-dev.apps.silver.devops.gov.bc.ca/api/doc#/traction-innkeeper/put_innkeeper_tenants__tenant_id__config
endpoint, you can update or set the fields in the arrays, but it doesn't seem to propagate that through if you set it to blank arrays.

See screenshot below request:

image

Then getting the tenant config after that shows the fields have not been updated:

image

### PR DESCRIPTION
**Pull request by shaangill025**
- resolve #607 
- resolve #608 
- resolve #667

### CODE REVIEW

**Review by loneil — CHANGES_REQUESTED** [review]
It's possible I'm really misunderstanding what the `self_issuer_permission` flag is to do here? Or there's some fixes needed.

I think the intention is if self_issuer_permission is **true** then a tenant that gets created can then connect to the endorser and register a DID without Innkeeper intervention right?

Trying locally on Docker, if I flip `self_issuer_permission` to true then create a new Tenant, I'm still seeing the blockage on the issuance switches (see below)
This might just be a minor Tenant UI code bug, see comments in code here.

image

Tried out switching up the conditional there locally, and the switches appeared as I would expect (again if I'm understanding the permission flag), however when trying to connect to the endorser it's still blocking the user:

image

If I'm a Tenant with `self_issuer_permission` true in my config I should be able to connect to the endorser without Innkeeper intervention right?


Then also, on the PR here (https://pr-737-tenant-ui-dev.apps.silver.devops.gov.bc.ca/innkeeper), where `self_issuer_permission` is **false**. I'm now unable to set the endorser and DID permissions as the Innkeeper.
It's expecting the `self_issuer_permission` field to be PUT back with the other config. Maybe this is just a matter of the Tenant UI sending back the current config value for that back in the PUT? If that's the intended flow? I know some other settings PUT commands around ACA-Py don't enforce including every field in the body and then it doesn't change it if it's not supplied (not sure if that's a consistent pattern or not)
I guess it could be nice in the event that self_issuer_permission is ON, that the Innkeeper could revoke that 🤷 So could have a switch for it as well, but that could be a future consideration. Need to not get the 422s for now.

image

**Comment by loneil** [comment]
Aside from comments above also a question probably for @esune for tracking future. Your idea for https://github.com/bcgov/traction/issues/607 was the endorser and did registration settings to be pulled as configured defaults (see 607 AC) for the endorser(s) and ledger(s).
I'm definitely content doing this `self_issuer_permission` boolean instead, but not sure if you want to keep https://github.com/bcgov/traction/issues/607 around for other future plans rather than close with this implementation? Maybe needs revisiting with future multi-ledger stuff (or just have this single `self_issuer_permission` be the permanent solution)

**Comment by loneil** [comment]
Unblocking PR review as I am out of service until Aug5 in case someone else needs to look at resolution of issues above and merge.

**Comment by esune** [comment]
Talked with @shaangill025 this morning: the `self_issuer_permission ` flag should only be used to drive the logic at the innkeeper level, without UI changes. If the `self_issuer_permission` is set to `true`, on tenant approval (manual or automatic) the settings enabling the tenant to connect to the endorser and publish their DID would be copied to the tenat's configuration, similarly to what happens when explicitly turning them on via UI/API.

File: plugins/traction_innkeeper/traction_innkeeper/v1_0/innkeeper/utils.py [hunk 1]
```diff
@@ -45,6 +45,11 @@ class TenantConfigSchema(OpenAPISchema):
         ),
         description="Public DID config",
     )
+    self_issuer_permission = fields.Bool(
+        required=False,
+        description="True if tenant can make itself issuer, false if only innkeeper can",
+        default=False,
```
Comment (esune) [inline plugins/traction_innkeeper/traction_innkeeper/v1_0/innkeeper/utils.py:51]: Might be better described as `True if tenants are automatically allowed to be issuers, False if innkeeper approval is required". Maybe the flag needs to be renamed as well to something more consistent to the behaviour, like `auto_issuer_permission`?

Comment (shaangill025) [reply plugins/traction_innkeeper/traction_innkeeper/v1_0/innkeeper/utils.py:48]: Changed to `auto_issuer` to be consistent with other flags in reservation config like `auto_approve`.

**Comment by loneil** [comment]
@shaangill025 unable to get a token as innkeeper tenant and seeing the following error in acapy logs.

```
return await handler(request)
File "/home/aries/.venv/lib/python3.9/site-packages/aries_cloudagent/admin/server.py", line 330, in check_token
return await handler(request)
File "/home/aries/.venv/lib/python3.9/site-packages/aries_cloudagent/admin/server.py", line 379, in check_multitenant_authorization
return await handler(request)
File "/home/aries/.venv/lib/python3.9/site-packages/aries_cloudagent/admin/server.py", line 442, in setup_context
return await task
File "/usr/local/lib/python3.9/asyncio/futures.py", line 284, in __await__
yield self # This tells Task to wait for completion.
File "/usr/local/lib/python3.9/asyncio/tasks.py", line 328, in __wakeup
future.result()
File "/usr/local/lib/python3.9/asyncio/futures.py", line 201, in result
raise self._exception
File "/usr/local/lib/python3.9/asyncio/tasks.py", line 256, in __step
result = coro.send(None)
File "/home/aries/traction_plugins/traction_innkeeper/v1_0/innkeeper/routes.py", line 88, in wrapper
raise err
File "/home/aries/traction_plugins/traction_innkeeper/v1_0/innkeeper/routes.py", line 72, in wrapper
ret = await func(request)
File "/home/aries/traction_plugins/traction_innkeeper/v1_0/innkeeper/routes.py", line 422, in tenant_create_token
tenant_record = await TenantRecord.retrieve_by_id(session, tenant_id)
File "/home/aries/.venv/lib/python3.9/site-packages/aries_cloudagent/messaging/models/base_record.py", line 243, in retrieve_by_id
return cls.from_storage(record_id, vals)
File "/home/aries/.venv/lib/python3.9/site-packages/aries_cloudagent/messaging/models/base_record.py", line 123, in from_storage
return cls(**params)
File "/home/aries/traction_plugins/traction_innkeeper/v1_0/innkeeper/models.py", line 266, in __init__
super().__init__(
TypeError: __init__() got an unexpected keyword argument 'self_issuer_permission'
```

**Comment by shaangill025** [comment]
@loneil All references to `self_issuer_permission` has been changed to `auto_issuer`.

**Comment by shaangill025** [comment]
Reopened as #748

**Pull request closed without merging**